### PR TITLE
[jit] add resolveType to Resolver

### PR DESCRIPTION
**Pull request by suo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #19239 [wip][jit] use fully qualified name for ScriptClasses
* #19238 [jit] Allow for segmented printing in PythonPrint
* **#19237 [jit] add resolveType to Resolver**
* #19236 [jit] Turn resolver into a class

Now that we have first-class types, various things will need to resolve
name strings into actual typeptrs. Currently we do this just be looking
up the name in our global registry, but once we add namespacing that will
no longer be sufficient.

This PR adds the infrastructure for resolving types, although the
functionality is the same as before.

Test Plan:
jit tests

gh-metadata: pytorch pytorch 19237 gh/suo/25/head

Differential Revision: [D14928289](https://our.internmc.facebook.com/intern/diff/D14928289)

### CODE REVIEW

File: torch/csrc/jit/script/resolver.h [hunk 1]
```diff
@@ -29,6 +29,9 @@ struct Resolver {
       const std::string& name,
       Function& m,
       const SourceRange& loc) const = 0;
+
+  // Resolve `name` to a TypePtr.
+  virtual TypePtr resolveType(const std::string& name) const = 0;
```
Comment (driazati) [inline torch/csrc/jit/script/resolver.h]: Once there are qualified names we need to be careful of `name` vs `qualname` vs `basename` to keep things clear

File: torch/csrc/jit/script/script_type_parser.h [hunk 2]
```diff
@@ -28,6 +33,7 @@ class TORCH_API ScriptTypeParser {
   at::TypePtr subscriptToType(
       const std::string& typeName,
       const Subscript& subscript) const;
+  ResolverPtr resolver_ = nullptr;
```
Comment (driazati) [inline torch/csrc/jit/script/script_type_parser.h]: Why would this ever need to be null?

**Review by driazati — APPROVED** [review]

File: torch/csrc/jit/script/script_type_parser.h (hunk 2 above)
Comment (suo) [reply torch/csrc/jit/script/script_type_parser.h]: If we ever wanted to use a "bare" ScriptTypeParser without an outer environment to look up type names in. Doesn't seem super likely, but there's no reason why we should require a resolver.

File: torch/csrc/jit/script/resolver.h (hunk 1 above)
Comment (zdevito) [reply torch/csrc/jit/script/resolver.h]: My vote is to use the QualName class when that change is made. Separate strings are easy to lose track of.

File: torch/csrc/jit/script/script_type_parser.h (hunk 2 above)
Comment (zdevito) [reply torch/csrc/jit/script/script_type_parser.h]: not used yet?

**Review by zdevito — APPROVED** [review]
Looks good! But it misses the addition of the resolveType override in the proxy object added in the last commit, which is problematic because it means that will be pretty likely to happen in the future as well.

**Comment by suo** [comment]
> Looks good! But it misses the addition of the resolveType override in the proxy object added in the last commit, which is problematic because it means that will be pretty likely to happen in the future as well.

There was indeed a small resolveType override [added](https://github.com/pytorch/pytorch/pull/19237/files#diff-c0a25c14600294e812cfbdd9ac3fcae2R2793) to MethodResolver. I think as long as we keep `Resolver` an interface-only thing, the proxy type isn't so bad as it will be forced to provide an implementation of whatever new method is added to Resolver. I will add a comment to that effect in Resolver

**Pull request closed without merging**